src/02-pa-upload.md

```diff
@@ -0,0 +1,112 @@
+# Upload
+
+[TODO: Add an illustration of this sub-protocol.]
+
+Uploading a report involves two requests to the leader. In the  *upload start
+request*, the client discovers the protocol-specific parameters and the endpoint
+URL of each helper. In the *upload finish request* uploads its report to the
+leader.
+
+## Upload Start
+
+**Request.**
+Let `[leader]` denote the URL of the leader's report ingestion endpoint. The
+client sends a POST request to `[leader]/upload_start` with the following
+message:
+
+```
+struct {
+  PATask task;
+  opaque id[16];
+} PAUploadStartReq;
+```
+
+The first field, `task` corresponds to the PA task for which a report will be
+generated. The second, `id` is a random, 32-byte string generated by the
+client.
+
+**Response.**
+The leader responds to well-formed requests to `[leader]/upload_start` with
+status 200 and the following message:
+
+```
+struct {
+  Url helper_urls<0..2^16-1>;
+  PAProto proto;
+  select (PAUploadStartResp.proto) {
+    case prio: PrioClientParam; // Section 3
+    case hits: HitsClientParam; // Section 4
+  }
+} PAUploadStartResp;
+```
+
+The message includes the URL of each helper and any protocol specific parameters
+the client needs to generate its report. The leader's response to malformed
+requests is specified in [the "Error handling"
+section](pa-error.md#common-abort-conditions).
+
+## Upload Finish
+
+For each URL `[helper]` in `PAClientParam.helper_urls`, the client sends a GET
+request to `[helper]/key_config`. The helper responds with status 200 and an
```

Comment:
I think the `key_config` request should be broken out into it's own sub-section.

Reply:
Ack, will do.